### PR TITLE
feat(websocket): use proxy for websocket connections

### PR DESCRIPTION
**Pull request by NewMeeh**
**Please describe the changes this PR makes and why it should be merged:**

The PR adds a way to use discord.js through a proxy.

**Status and versioning classification:**

I tested it and it works for my configuration. 
I am not sure how to update typings but I'll be happy to do so :) 
This PR adds one parameter but it is optional.
I will also be happy to add some documentation if needed.

### CODE REVIEW

**Comment by almeidx** [comment]
Documentation for this option needs to be added to:
- https://github.com/NewMeeh/discord.js/blob/ce08bca2e16537cf537c8e987db79d9fdc7e97f0/packages/discord.js/src/util/Options.js#L87
- https://github.com/NewMeeh/discord.js/blob/ce08bca2e16537cf537c8e987db79d9fdc7e97f0/packages/discord.js/typings/index.d.ts#L6779

**Review by didinele — CHANGES_REQUESTED** [review]
The key problem here is that your new agent option is not serializable over `worker_thread`s. This completely breaks the worker sharding strategy.

**Comment by NewMeeh** [comment]
@didinele I see the problem. I don't really know how threads works in Javascript, if there's no way to share the agent option between threads, should I construct the `agent` inside the `WebSocketShard` but this means adding a dependency 🤷‍♂️

**Comment by didinele** [comment]
I'm not confident we should go forward with this. Thoughts? @vladfrangu

**Comment by didinele** [comment]
Yeah. Frankly, if the agent API was universal and we found a nice interface for users to pass an instance even in worker mode, I'd be a lot more in favor of this, but as it is it feels whacky at best.

File: packages/ws/src/ws/WebSocketShard.ts [hunk 1]
```diff
@@ -185,8 +186,14 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
 
 		this.debug([`Connecting to ${url}`]);
 
+		const { proxy, httpsProxyAgentOptions } = this.strategy.options.proxyAgentOptions ?? {
+			proxy: undefined,
+			httpsProxyAgentOptions: undefined,
+		};
+		const agent = proxy ? new HttpsProxyAgent(proxy, httpsProxyAgentOptions) : undefined;
```
Comment (NotSugden) [inline packages/ws/src/ws/WebSocketShard.ts:193]:
```suggestion
		const { proxy, httpsProxyAgentOptions } = this.strategy.options.proxyAgentOptions ?? {
			proxy: undefined,
			httpsProxyAgentOptions: undefined,
		};
		const agent = proxy && new HttpsProxyAgent(proxy, httpsProxyAgentOptions);
```

**Comment by didinele** [comment]
Hi. Especially considering the direction we're taking of ensuring the package works in non-node envs, on top of everything else already discussed in this thread, we won't be going forward with this.

**Pull request closed without merging**